### PR TITLE
Relay parse assertion failures to debug console.

### LINKED ISSUE #648: No error information is displayed after parse error from assert
Trying to test the vscode extension with the PCAP DFDL schema.

I clobbered the magic number in the first bytes of an icmp1.cap file. This fails an assert at line 97 of pcap.dfdl.xsd, which is the first element of simple type in the schema. 

I step through with the debugger, and nothing displays *anything* about the assert failure. Execution ends. A little pop-up dialog says it wrote my infoset out, which it didn't. There's nothing showing that anything went wrong. All panels on the left are blank.

I was testing 1.3.0-RC2 version of the vscode extension.

### PR DESCRIPTION
**Pull request by arosien**
As part of previously-closed #648.

### CODE REVIEW

**Review by mbeckerle — APPROVED** [review]
I really cannot understand this code without further study of the whole cats effect approach.  My problem not yours, but my review is cursory as a result. I can't spot errors in this code.

File: server/core/src/main/scala/org.apache.daffodil.debugger.dap/Parse.scala [hunk 1]
```diff
@@ -99,25 +102,17 @@ object Parse {
             }
 
             val parse =
-              IO.interruptibleMany {
-                val parse_res = dp.parse(
+              IO.interruptibleMany(
```
[the remark this answers is not in the document]
Comment (arosien) [reply server/core/src/main/scala/org.apache.daffodil.debugger.dap/Parse.scala:105]: It's a rather niche method: https://typelevel.org/cats-effect/api/3.x/cats/effect/IO$.html#interruptibleMany[A](thunk:=%3EA):cats.effect.IO[A]. When wrapping a method that blocks a thread you need to decide if that computation is interruptible.

(This begs the question, does `parse` actually block the thread, waiting until the infoset? In my tests using `interruptibleMany` was the only way to kill a running parse.)

For these situations, you use `IO.blocking` if you don't care about interruption, `IO.interruptible` if you want the thread to be interrupted at most once, and `IO.interruptibleMany` if more than one thread interrupt should be attempted.

It's a rather big hammer, but the current choice seems to work without (me) knowing too much about the thread semantics of a parse.

File: server/core/src/main/scala/org.apache.daffodil.debugger.dap/Parse.scala [hunk 2]
```diff
@@ -99,25 +102,17 @@ object Parse {
             }
 
             val parse =
-              IO.interruptibleMany {
```
[the remark this answers is not in the document]
Comment (arosien) [reply server/core/src/main/scala/org.apache.daffodil.debugger.dap/Parse.scala:102]: I'll make an issue to document the magic number.

File: server/core/src/main/scala/org.apache.daffodil.debugger.dap/Parse.scala [hunk 3]
```diff
@@ -989,7 +985,7 @@ object Parse {
         schemaLocation: SchemaFileLocation,
         pointsOfUncertainty: List[PointOfUncertainty],
         delimiterStack: List[Delimiter],
-        diagnostics: List[Diagnostic]
+        diagnostics: List[org.apache.daffodil.api.Diagnostic]
```
[the remark this answers is not in the document]
Comment (arosien) [reply server/core/src/main/scala/org.apache.daffodil.debugger.dap/Parse.scala:988]: Correct: `PState` contains `api.Diagnostics`, whereas it seemed better in the DAP debugger backend to use `sapi.Diagnostics` for state.

**Comment by arosien** [comment]
> I really cannot understand this code without further study of the whole cats effect approach. My problem not yours, but my review is cursory as a result. I can't spot errors in this code.

I understand. I am currently working on a README to detail the backend architecture, which should help some of this.

For this particular bug though, it boils down to two concurrent computations going on in the `Parse` code:
- the ongoing Daffodil parse (`parsing`) which emits the infoset to some sink, while also pushing events into local queues; and
- a computation (`deliverParseData`) that transforms parse events into the current backend debugger state (stacktrace, etc.).
- 
What we want is that if there is an error during `parsing`, we should emit a DAP event with the error. However the previous code didn't do anything like that, and would incorrectly finalize the DAP event queue when the `deliverParseData` computation completed. (which in practice handled normal shutdown scenarios just fine)

Instead we now match when the error happened during `parsing`, enqueue the proper DAP event so the client will know about the problem, and then finalize the DAP event queue since we current shutdown the debugging session when such an error occurs.

**Review by scholarsmate — APPROVED** [review]
+1, but need to fix scalaFmt.

**Pull request merged**